## Импорт библиотек и данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Задания

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результаты каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

### Цели исследования

Необходимо провести анализ существующей базы данных книжного магазина и сформировать ценностное предложение для нового продукта.

In [14]:
query = 'SELECT * FROM books LIMIT 10'
pd.io.sql.read_sql(query, con=engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [15]:
query = 'SELECT * FROM authors LIMIT 10'
pd.io.sql.read_sql(query, con=engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [16]:
query = 'SELECT * FROM publishers LIMIT 10'
pd.io.sql.read_sql(query, con=engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [18]:
query = 'SELECT * FROM ratings LIMIT 10'
pd.io.sql.read_sql(query, con=engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [19]:
query = 'SELECT * FROM reviews LIMIT 10'
pd.io.sql.read_sql(query, con=engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [21]:
query = '''
--Выбираем книги выпущенные после 1го января 2000 года
SELECT COUNT(*) AS Количество_книг
FROM books 
WHERE publication_date > '2000-01-01'
'''

pd.io.sql.read_sql(query, con=engine)

Количество_книг
0              819

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [91]:
query = '''
SELECT books.title,
(SELECT COUNT(review_id) AS review_cnt --Считаем количество обзоров на книгу
    FROM reviews
    WHERE book_id = books.book_id),
(SELECT AVG(rating) --Считаем средний рейтинг на книгу
    FROM ratings
    WHERE book_id = books.book_id) AS avg_rating
FROM books, reviews
GROUP BY books.book_id
ORDER BY review_cnt DESC

'''

pd.io.sql.read_sql(query, con=engine)

title  review_cnt  avg_rating
0                              Twilight (Twilight  #1)           7    3.662500
1                                  Water for Elephants           6    3.977273
2                                     The Glass Castle           6    4.206897
3    Harry Potter and the Prisoner of Azkaban (Harr...           6    4.414634
4    The Curious Incident of the Dog in the Night-Time           6    4.081081
..                                                 ...         ...         ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel           0    3.666667
996                            The Natural Way to Draw           0    3.000000
997   The Cat in the Hat and Other Dr. Seuss Favorites           0    5.000000
998                          Essential Tales and Poems           0    4.000000
999  Disney's Beauty and the Beast (A Little Golden...           0    4.000000

[1000 rows x 3 columns]

Есть книги у которых очень мало обзоров, но при этом хороший рейтинг, надо будет их подсветить.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц


In [10]:
query = '''
SELECT p.publisher, COUNT(b.book_id)
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY COUNT(b.book_id) DESC
LIMIT 1
'''

pd.io.sql.read_sql(query, con=engine)

publisher  count
0  Penguin Books     42

Пенгвин это действительно знаменитое издательство

### Определите автора с самой высокой средней оценкой книг

In [ ]:
query = '''
SELECT a.author
FROM authors AS a
JOIN books AS b ON autho'''

In [92]:
query = '''
--
SELECT author, AVG(avg) AS average_rating --выводим автора, и среднюю оценку всех его книг
FROM(
    SELECT b.author_id, b.title, COUNT(r.rating), AVG(r.rating) --считаем кол-во рейтингов, средний рейтинг; выводим название книги и ID автора
    FROM books AS b
    JOIN ratings AS r ON r.book_id=b.book_id
    JOIN authors AS a ON a.author_id=b.author_id
    GROUP BY b.book_id
    HAVING COUNT(r.rating) > 50
    ) AS v1
JOIN authors AS a ON v1.author_id = a.author_id
GROUP BY author
ORDER BY average_rating DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con=engine)

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.283844

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [93]:
query = '''
        SELECT AVG(count) -- среднее количество обзоров на всех юзеров
        FROM(
            SELECT COUNT(review_id) -- среднее кол-во обзоров на каждого такого юзера
            FROM reviews as r
            WHERE username IN (
                SELECT username --юзеры с >50 обзоров
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 50
                )
            GROUP BY username
            ) AS v1
'''
pd.io.sql.read_sql(query, con=engine)

avg
0  24.333333

Пользователи которые ставят много оценок, так же пишут много обзоров.

## Вывод

Мы провели исследование базы данных. С 1 января 2000 года в ней появилось 819 новых книг. Наибольшее количество обзоров у книги Twilight(сумерки). Издательство выпустившее больше всего книг - Penguin Books. Автор с самыми высокорейтинговыми книгами - Джоан Роулинг. Самые активные пользователи в среднем оставляют 24 комментария.

Рекомендации: Подсветить книги с хорошим рейтингом но малым количеством отзывов, наверняка там есть hidden gems.